### ISSUE TITLE
[FG:InPlacePodVerticalScaling] Add additional E2E Burstable Pod QoS test cases for pods without resource limits 

### ISSUE THREAD

**Issue by jkyros** [kubernetes/kubernetes#124712]
### What would you like to be added?

It would be really great if  `InPlacePodVerticalScaling` could in-place resize cpu on containers that have requests only (no limits). 

Currently, `InPlacePodVerticalScaling` cannot successfully resize `cpu` against containers without limits.  e.g. given this pod: 
```
apiVersion: v1
kind: Pod
metadata:
  name: sleeper
  namespace: default
spec:
  containers:
  - args:
    - -c
    - sleep infinity
    command:
    - /bin/sh
    image: registry.k8s.io/ubuntu-slim:0.1
    imagePullPolicy: IfNotPresent
    name: sleeper
    resources:
      requests:
        cpu: 25m
        memory: 100Mi
```
If you change `cpu: 25m` to `cpu: 50m`, it will get stuck in `resize: InProgress` _forever_: 
```
  containerStatuses:
  - allocatedResources:
      cpu: 50m
      memory: 100Mi
    ...
    resources:
      requests:
        cpu: 25m
        memory: 100Mi
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2024-05-04T00:11:16Z"
  ...
  qosClass: Burstable
  resize: InProgress
```
### Why doesn't in-place work now without limits: 
- This looks like a result of this kubelet code [here](https://github.com/openshift/kubernetes/blob/258f1d5fb6491ba65fd8201c827e179432430627/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L556) not calculating the update if there are no limits (as in, if you do something naive like [this](https://github.com/kubernetes/kubernetes/compare/master...jkyros:kubernetes:minimal-patch-fix-inplacepodverticalscaling-limits) it works, but probably has other consequences and I don't have enough background here to propose that as a proper fix )
- Memory looks like it "works" without limits, but it's [special](https://github.com/openshift/kubernetes/blob/258f1d5fb6491ba65fd8201c827e179432430627/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L592) and...doesn't really

### What additional tests do we need: 

- The [current test suite](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/test/e2e/node/pod_resize.go#L664) includes some Burstable cases, but does not cover all corners, specifically: 
	- in-place increasing/reducing CPU requests on containers without limits 
	- in-place removing limits from a container that previously had them 
- Minimally could look something like this: https://github.com/kubernetes/kubernetes/commit/72953418202f7d412072b2259e996efc8e14abbd

### Additional background: 
- This was discussed in the `sig-node` meeting on Apr 9, 2024, where I was encouraged to inquire via Slack
- Per [that slack discussion](https://kubernetes.slack.com/archives/C06FSK01BGU/p1712688791899199) with @vinaykul  I was encouraged to open an issue to add additional test coverage so the "containers without limits" cases could be covered. 
- This isn't just "please do this for me", I'm willing to do work here to get these corners covered if need be, but I don't want to be "in the way"  :smile: 

/sig node 

### Why is this needed?

- Trying to use `InPlacePodVerticalScaling` in the Vertical Pod Autoscaler (VPA) [here](https://github.com/kubernetes/autoscaler/pull/6652) (which will probably also make #122836  happy). A bunch of the usage (and test suite) of the VPA is on containers without limits.

**Comment by dshebib** [comment]
My understanding is that the VPA does require both requests and limits, and the suggested use case is for QoS Guaranteed pods. If the Pod is Burstable, then it should already be able to scale its memory/cpu usage up to the specified limit or the node limit if none is specified without using the VPA. The problem seems to be that the VPA is trying to resize pods without a limit and leaving them InProgress, it doesn't seem like it should be touching those pods at all. So these tests should definitely be added, but perhaps the expectation should be that they aren't changed _and_ are not hung up in an InProgress resizing attempt.

As far as the reasons why it isn't ignoring pods that only have memory requests changed, the runtime manager will [refuse to resize pods without a memory limit](https://github.com/openshift/kubernetes/blob/258f1d5fb6491ba65fd8201c827e179432430627/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L727) but I'm not sure where it is attempting to change them in the first place.

**Comment by jkyros** [comment]
@dshebib This feels very "these aren't the droids you're looking for" Jedi-mind-trick-y -- like if you just tell me I don't need to use it then that somehow fixes it. It's not going to work.  :stuck_out_tongue: 

> My understanding is that the VPA does require both requests and limits, and the suggested use case is for QoS Guaranteed pods.

No, that is not correct. The VPA does not require both requests and limits, and the suggested use case is not limited to QoS Guaranteed pods. Where did you get this particular understanding? The VPA supports scaling of requests and/or limits. Here is a code entry point if you want to see what the recommender does: https://github.com/kubernetes/autoscaler/blob/12dba09f7dcb79df057a0ba4693dc3221d89b158/vertical-pod-autoscaler/pkg/admission-controller/resource/pod/recommendation/recommendation_provider.go#L71 

> The problem seems to be that the VPA is trying to resize pods without a limit and leaving them InProgress

Yes! We agree on this. The VPA wants to perform a completely reasonable operation, an operation that it would be performing normally via eviction, but that would be smoother for the user if it could do it in-place via `InPlacePodVerticalScaling`.  Ideally the VPA would do as much as it could in-place to avoid disrupting the targeted workload and fall back to eviction only when necessary. 

A consumer of `InPlacePodVerticalScaling` would not expect the sharp corner/surprise of it only working on pods with limits if there is not a meaningful technical reason why it should not, and getting stranded `InProgress` when it is known up front it can't ever succeed is especially surprising. The resources requests are important for scheduling and minimum guaranteed resources. Ideally they would be adjustable in-place. 

> perhaps the expectation should be that they aren't changed and are not hung up in an InProgress resizing attempt.

That's probably the _very least_ we should do here, but it makes `InPlacePodVerticalScaling` _so much_ less useful, it would be a tragedy to settle for that. Like I'd rather _be forced to write the code to make it work_ than enforce requiring limits. 

If we have "Very Good Reasons" for rejecting request-only resizes, we should state those technical reasons prominently and, as you say,  just reject the update so the user experience isn't painful, but I'm not seeing any "it can't work" here though so far -- I just see mostly "are you sure you need to do this?" (yes), and "It doesn't work right now" (I know, I don't like it  :stuck_out_tongue: ) 

>  it doesn't seem like it should be touching those pods at all

Can you...elaborate more...on why you believe this?  The scheduler uses the resource requests to decide where to put pods.  The resource request is the "low water mark"/minimum guaranteed resources for the container. The VPA absolutely should be (and needs to be) touching those pods it it wants pods to be scheduled properly and guaranteed at least the resources the VPA has recommended. If we have containers getting bigger, and we need them to keep those resources rather than overcommit, we want the scheduler to know that so it doesn't cram too many more pods on a node. And if we're using cluster autoscaler we'd want more nodes to scale up, etc 

>  the runtime manager will [refuse to resize pods without a memory limit](https://github.com/openshift/kubernetes/blob/258f1d5fb6491ba65fd8201c827e179432430627/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L727) but I'm not sure where it is attempting to change them in the first place.

That's the underlying cgroup resources though isn't it? Like am I wrong reading that call to `GetPodCgroupConfig` as "go get the existing cgroup values so we can compare them to our desired changes"? So that would be whatever is in the cgroup hierarchy, which is probably bad if it's nil, that makes sense.  

Because you can definitely have cgroups without memory limits ( well okay, I think technically the limit == max but it's effectively uncapped --  `memory.limit_in_bytes`/`memory.max` would just be a really big number or "max"). And I know that at the cgroup level the "requests" and "limits" definitely can be adjusted independently (not exactly the same down there, but you know what I mean). Which is why I linked to the code above where I naively made it work by just feeding it the "max" value that was already there rather than getting hung up on it: https://github.com/kubernetes/kubernetes/compare/master...jkyros:kubernetes:minimal-patch-fix-inplacepodverticalscaling-limits 

TL;DR: The VPA really does need "requests-only" scaling, but it's not just a niche case, there are really good general-case scheduling reasons this should work. I'd like to avoid "settling" if possible.

**Comment by dshebib** [comment]
@jkyros You're right, after reading the correct design doc [here](https://github.com/kubernetes/enhancements/tree/f6618a9a4fcfa98de4d4600743e4f2f6611c0497/keps/sig-node/1287-in-place-update-pod-resources), inplace resizing absolutely should be able to handle request-only resizes. Sorry for the confusion, I am new to the project and still learning.

As far as the test cases, looking at the current set of test cases , it looks like test cases which only change requests without limits are already added [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/pod_resize.go#L663)? . The problem as you have noted is that these tests are (passing)[https://testgrid.k8s.io/sig-node-presubmits#pr-kubelet-gce-cluster-e2e-inplace-pod-resize-containerd-main-v2] despite the issue you're having and which I'm trying to track down [pods getting stuck InProgress when only resizing memory requests](#125559).

This might be because container resources and pod allocations are checked, but the ContainerStatus.resources are [not checked](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/pod_resize.go#L1482) or because the resizeStatus itself is not checked. It's definitely strange (to me) that those other fields are updated properly but the resize isn't occuring properly.

**Comment by hshiina** [comment]
#126480 may fix this issue partially. It will allow resizing cpu and memory requests on containers whose limits are not configured.

Removing limits is not supported yet. I will try to implement it.
/assign

**Comment by maxcao13** [comment]
Seems like the underlying issue is fixed by https://github.com/kubernetes/kubernetes/pull/128718, but it doesn't actually implement most of the proposed test cases detailed here. Are the tests be instead to be resolved in @hshiina 's PR, and is that still planned to be merged?

**Comment by esotsal** [comment]
> Seems like the underlying issue is fixed by [#128718](https://github.com/kubernetes/kubernetes/pull/128718), but it doesn't actually implement most of the proposed test cases detailed here. Are the tests be instead to be resolved in [@hshiina](https://github.com/hshiina) 's PR, and is that still planned to be merged?

According to https://github.com/kubernetes/kubernetes/pull/127143#issuecomment-2702463416, I think plan is to have this in v1.34

**Comment by SergeyKanzhelev** [comment]
/remove-lifecycle stale

@natasha41575 is this still needed?

**Comment by natasha41575** [comment]
https://github.com/kubernetes/kubernetes/pull/128718 fixed the reported issue, and I checked and we do have some cases for resizing pods without limits. So no, I don't think this is needed.

I'll file a separate issue for allowing limits to be removed.

/close
